### PR TITLE
Add per-inode fscrypt_priv opaque field 

### PR DESCRIPTION
**Pull request by jtlayton**
This is still a very rough draft at this point, but should be enough for a PoC. The kernel client side of this is still a WIP.

This basically adds a 16-byte opaque field to inode_t and teaches the MDS to shlep it around alongside the size. The client will use this field to keep track of the "real size" of the inode, and it will always round up the traditional size field to the end of the last fscrypt block.

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Review by batrick — CHANGES_REQUESTED** [review]
Need to also encode here: https://github.com/ceph/ceph/blob/master/src/mds/CInode.cc#L1798

and add to the corresponding decode function

**Comment by jtlayton** [comment]
> Need to also encode here: https://github.com/ceph/ceph/blob/master/src/mds/CInode.cc#L1798
> 
> and add to the corresponding decode function

Thanks Patrick. SQUASH patch added on top that I'll fold into an earlier patch eventually.

File: src/mds/CInode.cc [hunk 1]
```diff
@@ -1800,6 +1800,7 @@ void CInode::encode_lock_ifile(bufferlist& bl)
       encode(get_inode()->truncate_size, bl); 
       encode(get_inode()->client_ranges, bl); 
       encode(get_inode()->inline_data, bl); 
+      encode_raw(get_inode()->fscrypt_priv, bl);
```
Comment (batrick) [inline src/mds/CInode.cc:1803]: A test case to ensure this works properly: changing the encrypted size but **not** changing the CephFS size correctly locks/unlocks the `ifile` lock. Presumably, the kernel won't be sending a redundant update to the size if it does not change but the encrypted size might change.

**Review by batrick — CHANGES_REQUESTED** [review]

File: src/mds/mdstypes.h [hunk 2]
```diff
@@ -590,6 +590,7 @@ struct inode_t {
   uint64_t   max_size_ever = 0; // max size the file has ever been
   uint32_t   truncate_seq = 0;
   uint64_t   truncate_size = 0, truncate_from = 0;
+  uint8_t    fscrypt_priv[CEPH_FSCRYPT_PRIVATE_SIZE] = {0};
```
[the remark this answers is not in the document]
Comment (jtlayton) [reply src/mds/mdstypes.h:593]: I think we're going to need to make this variable-length anyway, as I'm going to need to move the crypto context into this field as well.

**Comment by jtlayton** [comment]
For now I'm going to close out this PR as this is going to require some significant changes to incorporate the crypto context.

File: src/mds/CInode.cc [hunk 3]
```diff
@@ -1852,6 +1853,8 @@ void CInode::decode_lock_ifile(bufferlist::const_iterator& p)
       decode(_inode->truncate_size, p);
       decode(_inode->client_ranges, p);
       decode(_inode->inline_data, p);
+      if (struct_v >= 2)
+	decode_raw(_inode->fscrypt_priv, p);
```
[the remark this answers is not in the document]
Comment (jtlayton) [reply src/mds/CInode.cc:1857]: Thanks, will fix.

**Pull request closed without merging**